### PR TITLE
0.3.0 refactor

### PR DESCRIPTION
**Pull request by nshaaban-cPacket**

### CODE REVIEW

**Comment by nshaaban-cPacket** [comment]
@1c3t3a be critical! If you have any questions about design choices or anything else, please ask.

How is the changelog generated? Just want to make sure it generates with the correct changes.

File: .github/workflows/coverage.yml
Comment (1c3t3a) [inline .github/workflows/coverage.yml:46]: we could also think about committing the lock file. I already saw that with many other crates.

**Review by 1c3t3a — COMMENTED** [review]
Oke that's just for now! I plan to review the code tomorrow ;)

File: .github/workflows/coverage.yml
Comment (nshaaban-cPacket) [reply .github/workflows/coverage.yml:46]: A missing lock file would explain why the cache wasn't working.

Comment (1c3t3a) [reply .github/workflows/coverage.yml:46]: Yep, didn't think about it before. I just merged #114. Should we try to remove the lock file generation?

Comment (nshaaban-cPacket) [reply .github/workflows/coverage.yml]: > Yep, didn't think about it before. I just merged #114. Should we try to remove the lock file generation?

Woops, yup, you're right.

File: engineio/Cargo.toml
Comment (1c3t3a) [inline engineio/Cargo.toml:16]: Do we need rand here? The project actually compiles without it. Are there any other dependencies we could sort out?

File: .github/workflows/coverage.yml
Comment (nshaaban-cPacket) [reply .github/workflows/coverage.yml:5]: Wait until after this is merged, then yes. It's likely we'll want it around for the review process.

File: engineio/src/client/socket.rs
Comment (1c3t3a) [inline engineio/src/client/socket.rs:37]: Should we move this into a static variable? If we might change the version some time globally?

File: engineio/src/header.rs
Comment (1c3t3a) [inline engineio/src/header.rs:15]: How much would a usage of `&'a str` blow this up? Is it possible or just unnecessary because it produces a lot of `<'_>` bloat?

File: engineio/src/header.rs
Comment (1c3t3a) [inline engineio/src/header.rs:113]: How would this fail? Looks like an operation that always succeeds.

**Review by 1c3t3a — COMMENTED** [review]
Nice! That looks really good, thanks a lot @nshaaban-cPacket!! You really did a great job and refactored the codebase in a way that not enables a clean server implementation. Thanks.

File: engineio/benches/engineio.rs
Comment (1c3t3a) [inline engineio/benches/engineio.rs:7]: Would it make sense to call this in a ci job?

File: .github/workflows/coverage.yml
Comment (1c3t3a) [reply .github/workflows/coverage.yml:5]: Yep, that's true.

File: engineio/benches/engineio.rs
Comment (nshaaban-cPacket) [reply engineio/benches/engineio.rs:7]: Perhaps on tag creation? Takes a bit to run, wouldn't want it to block all PRs.

File: engineio/Cargo.toml
Comment (nshaaban-cPacket) [reply engineio/Cargo.toml:16]: Looks like rand is only used in socketio [here](https://github.com/1c3t3a/rust-socketio/blob/858d236b0337c15f8a9601355336e9b05f99167d/socketio/src/client/socket.rs#L4)

> Are there any other dependencies we could sort out?

https://github.com/1c3t3a/rust-socketio/pull/116

File: engineio/src/client/socket.rs
Comment (nshaaban-cPacket) [reply engineio/src/client/socket.rs:37]: https://github.com/1c3t3a/rust-socketio/pull/117

File: engineio/src/header.rs
Comment (nshaaban-cPacket) [reply engineio/src/header.rs:15]: Possibly a fairly large blow up and/or breakage as header names could have varying lifetimes, while the header map requires all the lifetimes of it's keys to be the same. Making a reference as a key to a map could lead to other strange behavior (like double references)

File: engineio/src/header.rs
Comment (nshaaban-cPacket) [reply engineio/src/header.rs:113]: https://github.com/1c3t3a/rust-socketio/pull/118

**Comment by nshaaban-cPacket** [comment]
CI had that intermittent failure again. PR set to draft until I have time to look at it, continue your review. https://github.com/1c3t3a/rust-socketio/runs/3602048511?check_suite_focus=true . @1c3t3a do you have any ideas why it might be failing?

File: engineio/src/header.rs
Comment (1c3t3a) [reply engineio/src/header.rs:15]: Ah I see, makes sense. Then nevermind ;)

File: engineio/benches/engineio.rs
Comment (1c3t3a) [reply engineio/benches/engineio.rs:7]: Ah yeah, tag creation sounds good! But lets first merge this PR, then we can merge it.

**Comment by nshaaban-cPacket** [comment]
> CI had that intermittent failure again. PR set to draft until I have time to look at it, continue your review. https://github.com/1c3t3a/rust-socketio/runs/3602048511?check_suite_focus=true . @1c3t3a do you have any ideas why it might be failing?

Solved in https://github.com/1c3t3a/rust-socketio/pull/119

**Comment by nshaaban-cPacket** [comment]
@1c3t3a I can't rebase 1c3t3a/refactoring. Not sure how you want to handle, I resolved all the merge conflicts by always taking refactoring's changes over main's (as the only changes to main were deprecation). Perhaps you could preform the rebase on 1c3t3a/refactoring (to ensure you have full visibility to the changes), then run tests, and compare with my branch where I did the refactor here https://github.com/nshaaban-cPacket/rust-socketio/tree/rebase-refactoring (made a [PR](https://github.com/1c3t3a/rust-socketio/pull/120) from that branch in case you wanted it).

**Comment by 1c3t3a** [comment]
> @1c3t3a I can't rebase 1c3t3a/refactoring. Not sure how you want to handle, I resolved all the merge conflicts by always taking refactoring's changes over main's (as the only changes to main were deprecation). Perhaps you could preform the rebase on 1c3t3a/refactoring (to ensure you have full visibility to the changes), then run tests, and compare with my branch where I did the refactor here https://github.com/nshaaban-cPacket/rust-socketio/tree/rebase-refactoring (made a [PR](https://github.com/1c3t3a/rust-socketio/pull/120) from that branch in case you wanted it).
> 
> > How is the changelog generated? Just want to make sure it generates with the correct changes.

Yep, I just rebased it, let's see what the ci says. For the changelog I use the tool [jilu](https://crates.io/crates/jilu). But it didn't always work in the past, so I'm usually writing it by hand.

**Comment by nshaaban-cPacket** [comment]
> > @1c3t3a I can't rebase 1c3t3a/refactoring. Not sure how you want to handle, I resolved all the merge conflicts by always taking refactoring's changes over main's (as the only changes to main were deprecation). Perhaps you could preform the rebase on 1c3t3a/refactoring (to ensure you have full visibility to the changes), then run tests, and compare with my branch where I did the refactor here https://github.com/nshaaban-cPacket/rust-socketio/tree/rebase-refactoring (made a [PR](https://github.com/1c3t3a/rust-socketio/pull/120) from that branch in case you wanted it).
> > > How is the changelog generated? Just want to make sure it generates with the correct changes.
> 
> Yep, I just rebased it, let's see what the ci says. For the changelog I use the tool [jilu](https://crates.io/crates/jilu). But it didn't always work in the past, so I'm usually writing it by hand.

Have you tried https://github.com/orhun/git-cliff ?

**Comment by 1c3t3a** [comment]
> Have you tried https://github.com/orhun/git-cliff ? I haven't personally, but stumbled across it yesterday.

Ah not yet, I might give it a try!

**Review by 1c3t3a — APPROVED** [review]

**Comment by 1c3t3a** [comment]
@nshaaban-cPacket Is there anything stopping this PR from getting merged? And what are our next steps? 
Probably:

- adopt Changelog
- Bump version and release it (alpha?)
- Create GitHub release as well

**Comment by nshaaban-cPacket** [comment]
> @nshaaban-cPacket Is there anything stopping this PR from getting merged? And what are our next steps?
> Probably:
> 
>     * adopt Changelog
> 
>     * Bump version and release it (alpha?)
> 
>     * Create GitHub release as well

Just tried to generate a changelog, far too many commits to make it readable. I'm going through now and fixing the commit messages, then I'll open a PR with the changelog.

Yup feel free to release as alpha, and do github releases.

**Comment by 1c3t3a** [comment]
Closed in favor of #122

**Pull request closed without merging**
